File: ProductOfArrayExceptSelf.js
var productExceptSelf = function(nums) {
    const leftProduct = new Array(nums.length);
    const rightProduct = new Array(nums.length);
    
    leftProduct[0] = nums[0];
    rightProduct[nums.length - 1] = nums[nums.length - 1];
    
    for (var i = 1; i < nums.length; i++) {
        leftProduct[i] = leftProduct[i - 1] * nums[i];
    }
    
     for (var i = nums.length - 2; i >= 0; i--) {
        rightProduct[i] = rightProduct[i + 1] * nums[i];
    }

    const result = new Array(nums.length);
    result[0] = rightProduct[1];
    
    for (var i = 1; i < nums.length - 1; i++) {
        result[i] = leftProduct[i - 1] * rightProduct[i + 1];
    }

    result[nums.length - 1] = leftProduct[nums.length - 2];
    
    return result;
};

File: ProductOfArrayExceptSelf.java
public class ProductOfArrayExceptSelf {
  public int[] productExceptSelf(int[] nums) {
    int[] leftProduct = new int[nums.length];  
    int leftAccProduct = 1;
    for (int i = 0; i < nums.length; i++) {
      leftAccProduct *= nums[i];
      leftProduct[i] = leftAccProduct;
    }

    int[] rightProduct = new int[nums.length];  
    int rightAccProduct = 1;
    for (int i = nums.length - 1; i >=0; i--) {
      rightAccProduct *= nums[i];
      rightProduct[i] = rightAccProduct;
    }

    int[] result = new int[nums.length];
    for (int i = 0; i < nums.length; i++) {
      if (i < 1) {
        result[i] = rightProduct[i + 1];
      } else if (i > nums.length - 2) {
        result[i] = leftProduct[i - 1];
      } else {
        result[i] = leftProduct[i - 1] * rightProduct[i + 1];
      }      
    }

    return result;
  }
}
